src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTracePipelineLogic.java

```diff
@@ -479,77 +497,131 @@ private boolean isLastNode(FlowNode flowNode) {
-                .build();
+            final StageData stageData = StageData.builder()
+                    .withName(pipelineNode.getName())
+                    .withStartTimeInMicros(pipelineNode.getStartTimeMicros())
+                    .withEndTimeInMicros(pipelineNode.getEndTimeMicros())
+                    .build();
 
-        stageBreakdownAction.put(stageData.getName(), stageData);
+            stageBreakdownAction.put(stageData.getName(), stageData);
+        } finally {
+            long end = DatadogAudit.currentTimeMillis();
+            DatadogAudit.log("DatadogTracePipelineLogic.updateStageBreakdown", start, end);
+        }
     }
 
     private void updateCIGlobalTags(Run run) {
-        final CIGlobalTagsAction ciGlobalTagsAction = run.getAction(CIGlobalTagsAction.class);
-        if(ciGlobalTagsAction == null) {
-            return;
+        long start = DatadogAudit.currentTimeMillis();
+        try {
+            final CIGlobalTagsAction ciGlobalTagsAction = run.getAction(CIGlobalTagsAction.class);
+            if(ciGlobalTagsAction == null) {
+                return;
+            }
+
+            final Map<String, String> tags = TagsUtil.convertTagsToMapSingleValues(DatadogUtilities.getTagsFromPipelineAction(run));
+            ciGlobalTagsAction.putAll(tags);
+        } finally {
+            long end = DatadogAudit.currentTimeMillis();
+            DatadogAudit.log("DatadogTracePipelineLogic.updateCIGlobalTags", start, end);
         }
+    }
 
-        final Map<String, String> tags = TagsUtil.convertTagsToMapSingleValues(DatadogUtilities.getTagsFromPipelineAction(run));
-        ciGlobalTagsAction.putAll(tags);
+    /**
+     * Creates a new Git client only if there is a Git information pending to calculate.
+     * This method tries to avoid creating Git clients as much as possible cause it's a very expensive operation.
+     * @param run
+     * @param pipelineNode
+     * @param node
+     * @param gitUrl
+     * @param gitCommit
+     * @return a git client if there is some git information to calculate. In other cases, it returns null.
+     */
+    private GitClient getGitClient(final Run<?, ?> run, final BuildPipelineNode pipelineNode, final FlowNode node, final String gitUrl, final String gitCommit) {
+        GitClient gitClient = null;
+        try {
+            if(!isValidCommit(gitCommit) && !isValidRepositoryURL(gitUrl)) {
+                return null;
+            }
+
+            GitCommitAction commitAction = run.getAction(GitCommitAction.class);
+            GitRepositoryAction repositoryAction = run.getAction(GitRepositoryAction.class);
+
+            final boolean commitInfoAlreadyCreated = commitAction != null && commitAction.getCommit() != null && commitAction.getCommit().equals(gitCommit);
+            final boolean repoInfoAlreadyCreated = repositoryAction != null && repositoryAction.getRepositoryURL() != null && repositoryAction.getRepositoryURL().equals(gitUrl);
```

Comment:
There boolean conditions are duplicated here and in `BuildData`. It might be worth extracting to a method.